sgx/trusted/src/wrapper/sgx_retrieve_key_impl.c

```diff
@@ -6,11 +6,203 @@
+    return 1;
+  }
+
+  if (message.response_header->batch_count != 1)
+  {
+    // kmyth_sgx_log(LOG_ERR, "Expected to receive one response; received: %d",
+    //           message.response_header->batch_count);
+    kmip_free_response_message(ctx, &message);
+    kmip_set_buffer(ctx, NULL, 0);
+    return 1;
+  }
+
+  ResponseBatchItem batch_item = message.batch_items[0];
+
+  if (batch_item.operation != KMIP_OP_GET)
+  {
+    kmyth_sgx_log(LOG_ERR, "Did not receive a KMIP Get response.");
+    kmip_free_response_message(ctx, &message);
+    kmip_set_buffer(ctx, NULL, 0);
+    return 1;
+  }
+  if (batch_item.result_status != KMIP_STATUS_SUCCESS)
+  {
+    // kmyth_sgx_log(LOG_ERR, "The KMIP Get request failed: %.*s",
+    //           batch_item.result_message->size,
+    //           batch_item.result_message->value);
+    kmip_free_response_message(ctx, &message);
+    kmip_set_buffer(ctx, NULL, 0);
+    return 1;
+  }
+
+  GetResponsePayload *payload =
+    (GetResponsePayload *) batch_item.response_payload;
+  if (payload->object_type != KMIP_OBJTYPE_SYMMETRIC_KEY)
+  {
+    kmyth_sgx_log(LOG_ERR, "The received KMIP object is not a symmetric key.");
+    kmip_free_response_message(ctx, &message);
+    kmip_set_buffer(ctx, NULL, 0);
+    return 1;
+  }
+
+  SymmetricKey *symmetric_key = (SymmetricKey *) payload->object;
+  KeyBlock *key_block = symmetric_key->key_block;
+  KeyValue *key_value = key_block->key_value;
+  ByteString *key_material = key_value->key_material;
+
+  // Set up the official ID and key buffers and clean up.
+  *id = calloc(payload->unique_identifier->size, sizeof(unsigned char));
+  if (*id == NULL)
+  {
+    kmyth_sgx_log(LOG_ERR, "Failed to allocate the ID buffer.");
+    kmip_free_response_message(ctx, &message);
+    kmip_set_buffer(ctx, NULL, 0);
+    return 1;
+  }
+  *id_len = payload->unique_identifier->size;
+  memcpy(*id, payload->unique_identifier->value, *id_len);
+
+  *key = calloc(key_material->size, sizeof(unsigned char));
+  if (*key == NULL)
+  {
+    kmyth_sgx_log(LOG_ERR, "Failed to allocate the key buffer.");
+    kmyth_enclave_clear_and_free(*id, *id_len);
+    *id = NULL;
+    *id_len = 0;
+    kmip_free_response_message(ctx, &message);
+    kmip_set_buffer(ctx, NULL, 0);
+    return 1;
+  }
+  *key_len = key_material->size;
+  memcpy(*key, key_material->value, *key_len);
+
+  kmip_free_response_message(ctx, &message);
+  kmip_set_buffer(ctx, NULL, 0);
+
+  return 0;
+}
+
 //############################################################################
 // enclave_retrieve_key()
 //############################################################################
-int enclave_retrieve_key(EVP_PKEY * enclave_sign_privkey, X509 * peer_cert)
+int enclave_retrieve_key(EVP_PKEY * enclave_sign_privkey, X509 * peer_cert,
+                         const char *server_host, int server_host_len,
+                         const char *server_port, int server_port_len)
 {
+  int ret_val, ret_ocall;
```

Comment:
The variable ret_ocall should be of type sgx_status_t.

Reply:
Okay, I'll fix that.